GUIDELINES.md

```diff
@@ -0,0 +1,159 @@
+# Technical and working style and guidelines
+
+*Challenging this documentation is encouraged. Please set up an issue for challenges, additions or corrections. For explanations, please consult the `dev` Slack channel*
+
+In this document we set out some basic "tips" (either mandatory or encouraged) which should guide your way of working, regardless of the task.
+
+* [Foreword](#foreword)
+* [`python is not pandas`](#python-is-not-pandas)
+* [Design patterns](#design-patterns)
+* [Programming](#programming)
+* [Critical thinking](#critical-thinking)
+* [Naming conventions](#naming-conventions)
+* [Spaces and spacing](#spaces-and-spacing)
+* [Comments and docs](#comments-and-docs)
+
+## Foreword
+
+In advance, we recommend installing the autoformatter [black](https://pypi.org/project/black/) in your IDE. In future we will use `flake8` (or similar) for automatically checking our python codebases. 
+
+Hopefully the conventions laid out here are the easy and intuitive set of [pep8](https://www.python.org/dev/peps/pep-0008/). 
+
+Code reviewers: it is on you to ensure that this style guide has been followed: there are no points for being lenient, but there [non-redeemable] points for being opinionated! We should all feel pressured into making sure that our code meets an acceptable standard.
+
+## `python is not pandas`
+
+tldr; Using `pandas` as a means to perform transformations or calculations on your data should be avoided, unless it clearly simplifies the logic and readability of your code. That is not so say that you should not use `pandas`, but rather that you justify to yourself that `pandas` isn't making your life harder in lieu of using standard `python` tools.
+
+Code containing lots of `pandas` operations are almost impossible to review, and therefore have the capacity to accumulate vast numbers of bugs.
+
+* In general, `pandas` makes column-wise operations and IO (reading/writing files) dead easy. That said, `pandas` column-wise operations are inherited from `numpy`, and `numpy` is generally accepted in the place of dataframes.
+* `pandas` is enormous, in many ways. If it can be omitted from your code then you can make big savings in terms of memory usage and requirements clashes, and even CPU time.
+* Instead of Googling how to achieve something in `pandas` with an obscure chaining of functions, break the problem down and solve it yourself. It is highly unlikely that the `pandas` approach to reshaping your data will beat using tools from `numpy`, `itertools`, `functools` and `toolz`, even if you switch to representing data in `numpy` arrays or even as `list` of `dict` (`[{'value': 21}, {'value': 45}]`).
+
```

Comment:
Add something reassuring here and encourage people to ask for advice rather than struggle to convert their `pandas` code? I imagine it will be a hard habit to break in some cases.

In the long run it could be good to create a set of examples of code that could be modularised into `numpy` or `python` based functions and those which are sensible to leave as `pandas`.

Reply:
It should be noted here that this logic goes against a lot of what data scientists hear when they first begin learning - "Don't write the code for anything that someone else has already sorted out". Might not always be obvious to everyone when to break this rule or not.

Reply:
thanks, I've added a couple of notes to this effect